decisions/0004-sync-routes-from-k8s-to-cc.md

```diff
@@ -0,0 +1,113 @@
+1. [nsync-bulker](https://github.com/cloudfoundry-attic/nsync/tree/master/cmd/nsync-bulker)
+1. [ProcessesSync](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/diego/processes_sync.rb)
+1. [Copilot::Sync](https://github.com/cloudfoundry/cloud_controller_ng/blob/master/lib/cloud_controller/copilot/sync.rb)
+1. [hm9000](https://github.com/cloudfoundry-attic/hm9000)
+
+Each of these components share a similar goal: fetch desired state and ensure it is consistently represented everywhere where it needs to become actual state.
+In practice, the modern versions of these bulk sync loops page through CCDB or the CF API and make requests to external components (diego, copilot, metac) to ensure that their internal states are consistent with values in CCDB.
+
+### Kubernetes
+
+Kubernetes is also implemented via a series of "bulk sync loops" that are more commonly known as controllers, but there are some key differences in how these convergence loops work.
+1. k8s has a ["watch"](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes) API so clients clients can subscribe to incremental changes in desired and actual state
+1. [k8s.io/client-go/cache](https://godoc.org/k8s.io/client-go/tools/cache) provides a mechanism for clients to build and maintain local caches of kubernetes state
+1. [kubebuilder](https://book.kubebuilder.io/) Provides, amongst other things, an SDK for building reconciliation loops on top of watches and client side caching
+
+### Problems and Solutions
+
+To illustrate the necessity of these sync loops, it helps to think about how they improve resiliency in 2 cases of temporary network partition from the perspective of the CF API: 
+1. Problem: Updates to CCDB succeed ✅, but updates to the Route CRD fail ❌
+1. Problem: Updates to the Route CRD succeed ✅, but updates to CCDB fail ❌
+
+To handle these 2 failure modes, several approaches can be taken. 
+
+##### Mimic historical sync loops:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, but allow the CCDB transaction to commit. A bulk sync loop will detect that the Route CRD is "missing" and create it later.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution A: Raise the error to the user, and don't do the Route CRD write until CCDB changes are committed.
+   1. Solution B: A bulk sync loop will detect that the Route CRD is "extraneous" and delete it later.
+
+##### Forgo a bulk sync and attempt to bootstrap consistency with k8s off CCDB transactions:
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
+   1. Solution: Roll back changes to the Route CRD... (but what if that still fails due to network partition?!)
+
+##### Sync from k8s to CCDB
+1. Problem: CCDB ✅ & Route CRD ❌
+   1. Solution: Raise the error to the user, abort the CCDB transaction.
+1. Problem: Route CRD ✅ & CCDB ❌
```

Comment:
Would line 54 mean the route CRD update would need to be inside the CCDB transaction? And if the route CRD update succeeds, we let the CCDB transaction go through? If the route CRD update is inside the CCDB transaction would we be able to run into the problem in line 55?

Reply:
> Would line 54 mean the route CRD update would need to be inside the CCDB transaction?

yes

> And if the route CRD update succeeds, we let the CCDB transaction go through?

yes

> If the route CRD update is inside the CCDB transaction would we be able to run into the problem in line 55?

yes, if the route CRD update is inside the CCDB transaction, it's possible for the CRD create/update/delete to succeed, but the transaction commit to fail. In that event we have a Route CRD existing in correct form in k8s and we could rely on our reconciliation logic to eventually update the CF API to match.